File: Leetcode_重刷紀錄/0014.Problems643. Maximum Average Subarray I/MaximumAverageSubarrayI.java
//1679. Max Number of K-Sum Pairs
class Solution {
    public double findMaxAverage(int[] nums, int k) {
        int vectorSize = nums.length;
        int i = 0;
        double maxSum = 0.0, curSum = 0.0;

        do {
            maxSum += nums[i];
        }while(++i < k);
        // now i == k
        curSum = maxSum;

        // 会有 i == k == vectorSize 的情况
        // 所以不能使用 do while
        while (i < vectorSize) {
            curSum += nums[i] - nums[i - k];
            if (maxSum < curSum) {
                maxSum = curSum;
            }
            i++;
        }

        return maxSum / k;
    }
}

File: Leetcode_重刷紀錄/0014.Problems643. Maximum Average Subarray I/ MaximumAverageSubarrayI.py
# 1679. Max Number of K-Sum Pairs
class Solution:
    def findMaxAverage(self, nums: List[int], k: int) -> float:
        vectorSize = len(nums)
        i = 0
        maxSum = 0.0
        curSum = 0.0

        for i in range(k):
            maxSum += nums[i]
        i = k
        curSum = maxSum

        while i < vectorSize:
            curSum += nums[i] - nums[i - k]
            if maxSum < curSum:
                maxSum = curSum
            i += 1

        return maxSum / k

# 別人給的解法，速度差不多 稍微慢一點點

class Solution:
    def findMaxAverage(self, nums: List[int], k: int) -> float:
        
        # Initialize currSum and maxSum to the sum of the initial k elements
        currSum = maxSum = sum(nums[:k])

        # Start the loop from the kth element 
        # Iterate until you reach the end
        for i in range(k, len(nums)):

            # Subtract the left element of the window
            # Add the right element of the window
            currSum += nums[i] - nums[i - k]
            
            # Update the max
            maxSum = max(maxSum, currSum)

        # Since the problem requires average, we return the average
        return maxSum / k